File: frontend/src/lib/calculator.ts
/**
 * Pure calculation engine for badminton court session expense splits.
 * Clean architecture domain logic with zero framework or I/O dependencies.
 */

export type SplitMode = 'even' | 'fixed_female_discount' | 'fixed_female' | 'multi_stage'

export interface EarlyLeaverConfig {
  count: number
  stage1Ratio: number // e.g. 0.5 for 1 hr out of 2 hr session
  stage1Shuttlecocks?: number
  stage1ShuttleFee?: number
}

export interface CalculatorInputs {
  courtFee: number
  shuttlecockCount: number
  shuttlecockUnitPrice: number
  shuttlecockFee?: number // total override if provided, else count * unitPrice
  maleCount: number
  femaleCount: number
  splitMode: SplitMode
  femaleDiscount?: number // VND, e.g. 10000
  fixedFemaleFee?: number // VND, e.g. 30000
  earlyLeaverConfig?: EarlyLeaverConfig
}

export interface CalculationResult {
  totalExpenses: number
  courtFee: number
  shuttlecockFee: number
  totalParticipants: number
  maleCount: number
  femaleCount: number
  splitMode: SplitMode
  maleFee: number
  femaleFee: number
  earlyFee?: number
  stayFee?: number
  earlyCount?: number
  stayCount?: number
  totalCollected: number
  fundBuffer: number // totalCollected - totalExpenses
}

/**
 * Rounds a VND monetary amount up to the nearest multiple of 1.000 VND.
 * e.g. 35100 -> 36000, 35000 -> 35000, 0 -> 0.
 */
export function roundUpTo1000(amount: number): number {
  if (amount <= 0 || !Number.isFinite(amount)) {
    return 0
  }
  return Math.ceil(amount / 1000) * 1000
}

/**
 * Calculates session splits based on configured rules and expense inputs.
 */
export function calculateSessionSplit(inputs: CalculatorInputs): CalculationResult {
  const courtFee = Math.max(0, inputs.courtFee || 0)
  const shuttlecockFee =
    inputs.shuttlecockFee !== undefined && inputs.shuttlecockFee >= 0
      ? inputs.shuttlecockFee
      : Math.max(0, inputs.shuttlecockCount || 0) * Math.max(0, inputs.shuttlecockUnitPrice || 0)

  const totalExpenses = courtFee + shuttlecockFee
  const maleCount = Math.max(0, Math.floor(inputs.maleCount || 0))
  const femaleCount = Math.max(0, Math.floor(inputs.femaleCount || 0))
  const totalParticipants = maleCount + femaleCount

  if (totalParticipants === 0 || totalExpenses === 0) {
    return {
      totalExpenses,
      courtFee,
      shuttlecockFee,
      totalParticipants,
      maleCount,
      femaleCount,
      splitMode: inputs.splitMode,
      maleFee: 0,
      femaleFee: 0,
      totalCollected: 0,
      fundBuffer: 0,
    }
  }

  let maleFee = 0
  let femaleFee = 0
  let earlyFee: number | undefined
  let stayFee: number | undefined
  let earlyCount: number | undefined
  let stayCount: number | undefined
  let totalCollected = 0

  switch (inputs.splitMode) {
    case 'even': {
      const rawFee = totalExpenses / totalParticipants
      const fee = roundUpTo1000(rawFee)
      maleFee = maleCount > 0 ? fee : 0
      femaleFee = femaleCount > 0 ? fee : 0
      totalCollected = maleCount * maleFee + femaleCount * femaleFee
      break
    }

    case 'fixed_female_discount': {
      const discount = Math.max(0, inputs.femaleDiscount || 0)
      if (femaleCount === 0 || discount === 0) {
        // Fallback to even split if no females or 0 discount
        const fee = roundUpTo1000(totalExpenses / totalParticipants)
        maleFee = maleCount > 0 ? fee : 0
        femaleFee = femaleCount > 0 ? fee : 0
      } else if (maleCount === 0) {
        // Only female players
        const fee = roundUpTo1000(totalExpenses / femaleCount)
        maleFee = 0
        femaleFee = fee
      } else {
        const rawFemaleBase = (totalExpenses - femaleCount * discount) / totalParticipants
        femaleFee = Math.max(0, roundUpTo1000(rawFemaleBase))
        maleFee = femaleFee + discount
      }
      totalCollected = maleCount * maleFee + femaleCount * femaleFee
      break
    }

    case 'fixed_female': {
      const fixedFee = Math.max(0, inputs.fixedFemaleFee || 0)
      if (femaleCount === 0) {
        const fee = roundUpTo1000(totalExpenses / maleCount)
        maleFee = fee
        femaleFee = 0
      } else if (maleCount === 0) {
        maleFee = 0
        femaleFee = fixedFee > 0 ? fixedFee : roundUpTo1000(totalExpenses / femaleCount)
      } else {
        femaleFee = fixedFee
        const remainingExpenses = Math.max(0, totalExpenses - femaleCount * fixedFee)
        maleFee = roundUpTo1000(remainingExpenses / maleCount)
      }
      totalCollected = maleCount * maleFee + femaleCount * femaleFee
      break
    }

    case 'multi_stage': {
      const earlyConfig = inputs.earlyLeaverConfig || { count: 0, stage1Ratio: 0.5 }
      earlyCount = Math.min(totalParticipants - 1, Math.max(0, Math.floor(earlyConfig.count || 0)))
      stayCount = totalParticipants - earlyCount

      if (earlyCount === 0 || stayCount === 0) {
        const fee = roundUpTo1000(totalExpenses / totalParticipants)
        maleFee = maleCount > 0 ? fee : 0
        femaleFee = femaleCount > 0 ? fee : 0
        earlyFee = fee
        stayFee = fee
      } else {
        const ratio1 = Math.min(1, Math.max(0, earlyConfig.stage1Ratio || 0.5))
        const stage1CourtCost = courtFee * ratio1

        const stage1ShuttleCost =
          earlyConfig.stage1ShuttleFee !== undefined
            ? earlyConfig.stage1ShuttleFee
            : earlyConfig.stage1Shuttlecocks !== undefined
              ? earlyConfig.stage1Shuttlecocks * Math.max(0, inputs.shuttlecockUnitPrice || 0)
              : shuttlecockFee * ratio1

        const cost1 = stage1CourtCost + stage1ShuttleCost
        const cost2 = Math.max(0, totalExpenses - cost1)

        earlyFee = roundUpTo1000(cost1 / totalParticipants)
        stayFee = earlyFee + roundUpTo1000(cost2 / stayCount)

        // For display parity: male/female fees reflect stayer and early fees
        maleFee = stayFee
        femaleFee = stayFee
      }

      totalCollected = (earlyCount || 0) * (earlyFee || 0) + (stayCount || 0) * (stayFee || 0)
      break
    }
  }

  const fundBuffer = totalCollected - totalExpenses

  return {
    totalExpenses,
    courtFee,
    shuttlecockFee,
    totalParticipants,
    maleCount,
    femaleCount,
    splitMode: inputs.splitMode,
    maleFee,
    femaleFee,
    earlyFee,
    stayFee,
    earlyCount,
    stayCount,
    totalCollected,
    fundBuffer,
  }
}

File: backend/app/services/calculator.py
"""
Pure calculation service for badminton court session splits.
Clean architecture domain logic with zero framework or database dependencies.
"""

import math

from backend.app.schemas.session import (
    SessionCalculationRequest,
    SessionCalculationResponse,
)


def round_up_to_1000(amount: float) -> float:
    """
    Rounds a VND monetary amount up to the nearest multiple of 1.000 VND.
    e.g. 35100.0 -> 36000.0, 35000.0 -> 35000.0, 0.0 -> 0.0.
    """
    if amount <= 0 or not math.isfinite(amount):
        return 0.0
    return float(math.ceil(amount / 1000.0) * 1000.0)


def calculate_session_split(req: SessionCalculationRequest) -> SessionCalculationResponse:
    """
    Calculates per player fees, total expenses, collected amount, and buffer.
    """
    court_fee = max(0.0, float(req.court_fee))
    if req.shuttlecock_fee is not None and req.shuttlecock_fee >= 0:
        shuttlecock_fee = float(req.shuttlecock_fee)
    else:
        unit_price = max(0.0, float(req.shuttlecock_unit_price))
        shuttlecock_fee = max(0.0, float(req.shuttlecock_count)) * unit_price

    total_expenses = court_fee + shuttlecock_fee
    male_count = max(0, int(req.male_count))
    female_count = max(0, int(req.female_count))
    total_participants = male_count + female_count

    if total_participants == 0 or total_expenses == 0.0:
        return SessionCalculationResponse(
            total_expenses=total_expenses,
            court_fee=court_fee,
            shuttlecock_fee=shuttlecock_fee,
            total_participants=total_participants,
            male_count=male_count,
            female_count=female_count,
            split_mode=req.split_mode,
            male_fee=0.0,
            female_fee=0.0,
            early_fee=None,
            stay_fee=None,
            early_count=None,
            stay_count=None,
            total_collected=0.0,
            fund_buffer=0.0,
        )

    male_fee = 0.0
    female_fee = 0.0
    early_fee: float | None = None
    stay_fee: float | None = None
    early_count: int | None = None
    stay_count: int | None = None

    if req.split_mode == "even":
        raw_fee = total_expenses / total_participants
        fee = round_up_to_1000(raw_fee)
        male_fee = fee if male_count > 0 else 0.0
        female_fee = fee if female_count > 0 else 0.0
        total_collected = (male_count * male_fee) + (female_count * female_fee)

    elif req.split_mode == "fixed_female_discount":
        discount = max(0.0, float(req.female_discount or 0.0))
        if female_count == 0 or discount == 0.0:
            fee = round_up_to_1000(total_expenses / total_participants)
            male_fee = fee if male_count > 0 else 0.0
            female_fee = fee if female_count > 0 else 0.0
        elif male_count == 0:
            fee = round_up_to_1000(total_expenses / female_count)
            male_fee = 0.0
            female_fee = fee
        else:
            raw_female_base = (total_expenses - (female_count * discount)) / total_participants
            female_fee = max(0.0, round_up_to_1000(raw_female_base))
            male_fee = female_fee + discount
        total_collected = (male_count * male_fee) + (female_count * female_fee)

    elif req.split_mode == "fixed_female":
        fixed_fee = max(0.0, float(req.fixed_female_fee or 0.0))
        if female_count == 0:
            male_fee = round_up_to_1000(total_expenses / male_count)
            female_fee = 0.0
        elif male_count == 0:
            male_fee = 0.0
            if fixed_fee > 0.0:
                female_fee = fixed_fee
            else:
                female_fee = round_up_to_1000(total_expenses / female_count)
        else:
            female_fee = fixed_fee
            remaining_expenses = max(0.0, total_expenses - (female_count * fixed_fee))
            male_fee = round_up_to_1000(remaining_expenses / male_count)
        total_collected = (male_count * male_fee) + (female_count * female_fee)

    elif req.split_mode == "multi_stage":
        early_config = req.early_leaver_config
        early_cnt = max(0, int(early_config.count)) if early_config else 0
        early_count = min(total_participants - 1, early_cnt)
        stay_count = total_participants - early_count

        if early_count == 0 or stay_count == 0:
            fee = round_up_to_1000(total_expenses / total_participants)
            male_fee = fee if male_count > 0 else 0.0
            female_fee = fee if female_count > 0 else 0.0
            early_fee = fee
            stay_fee = fee
        else:
            ratio1 = min(1.0, max(0.0, float(early_config.stage1_ratio if early_config else 0.5)))
            stage1_court_cost = court_fee * ratio1

            if early_config and early_config.stage1_shuttle_fee is not None:
                stage1_shuttle_cost = float(early_config.stage1_shuttle_fee)
            elif early_config and early_config.stage1_shuttlecocks is not None:
                u_price = max(0.0, float(req.shuttlecock_unit_price))
                stage1_shuttle_cost = float(early_config.stage1_shuttlecocks) * u_price
            else:
                stage1_shuttle_cost = shuttlecock_fee * ratio1

            cost1 = stage1_court_cost + stage1_shuttle_cost
            cost2 = max(0.0, total_expenses - cost1)

            early_fee = round_up_to_1000(cost1 / total_participants)
            stay_fee = early_fee + round_up_to_1000(cost2 / stay_count)

            male_fee = stay_fee
            female_fee = stay_fee

        total_collected = (early_count * (early_fee or 0.0)) + (stay_count * (stay_fee or 0.0))

    else:
        # Fallback
        fee = round_up_to_1000(total_expenses / total_participants)
        male_fee = fee if male_count > 0 else 0.0
        female_fee = fee if female_count > 0 else 0.0
        total_collected = (male_count * male_fee) + (female_count * female_fee)

    fund_buffer = total_collected - total_expenses

    return SessionCalculationResponse(
        total_expenses=total_expenses,
        court_fee=court_fee,
        shuttlecock_fee=shuttlecock_fee,
        total_participants=total_participants,
        male_count=male_count,
        female_count=female_count,
        split_mode=req.split_mode,
        male_fee=male_fee,
        female_fee=female_fee,
        early_fee=early_fee,
        stay_fee=stay_fee,
        early_count=early_count,
        stay_count=stay_count,
        total_collected=total_collected,
        fund_buffer=fund_buffer,
    )
